### ISSUE TITLE
Can not import_playbook from collection

### ISSUE THREAD

**Issue by provenvelocity** [ansible/ansible#74917]
### Summary

When I use a collection and try to reference a playbook in the collection I get an error. 

ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/nginxinc.nginx_core.playbooks.deploy-nginx.yml' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option

### Issue Type

Bug Report

### Component Name

import_playbook _collection_finder.py

### Ansible Version

I tested in ansible [core 2.11.0] & Devel with the same results but currently, I'm testing what my users will use which was 2.11.1

```
Testing in this though
ansible [core 2.11.1] 
  config file = /Users/H217689/kp/ansible/provenvelocity/devops_test/ansible.cfg
  configured module search path = ['/Users/H217689/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/H217689/kp/ansible/provenvelocity/pv_devops_venv/lib/python3.9/site-packages/ansible
  ansible collection location = /Users/H217689/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/H217689/kp/ansible/provenvelocity/pv_devops_venv/bin/ansible
  python version = 3.9.5 (default, May  4 2021, 03:36:27) [Clang 12.0.0 (clang-1200.0.32.29)]
  jinja version = 3.0.1
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed

Nill
```

### OS / Environment

Mac testing on centos/8

### Steps to Reproduce

Download my testing repo and run the steps and command in the import playbook issue portion of the readme.

https://github.com/provenvelocity/devops_test

Also I am not sure how any of the tests you have which are many pass.... :)

### Expected Results

I expect it to run the playbook from nginx.... I am using nginx as an example of the issue I have much more complicated internal playbooks that I wish to run. 

### Issues in code I found 

Starts in this code:

https://github.com/ansible/ansible/blob/2cbfd1e350cbe1ca195d33306b5a9628667ddda8/lib/ansible/playbook/playbook_include.py

_get_collection_playbook_path

But the code below is the issue. 

https://github.com/ansible/ansible/blob/baa371e7b5f9e207cd9aa5967d48812e2fb13aec/lib/ansible/utils/collection_loader/_collection_finder.py

This code seems to be the issue. The code below is never called with a ref_type which if it ever was would cause a recursion issue because the try_parse_fqcr ends up calling is_valid_fqcr when all is said and done so it loops until it crashes if you passed ref_type.  The code that is called also does not return true for the passed ref.  

Specific line with issue:
return bool(re.match(AnsibleCollectionRef.VALID_FQCR_RE, ref))

### Example print debugging hacking:

_get_collection_playbook_path
&nbsp;&nbsp;try_parse_fqcr
&nbsp;&nbsp;&nbsp;&nbsp;from_fqcr
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;print ref "nginxinc.nginx_core.playbooks.deploy-nginx.yml"
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;is_valid_fqcr
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return bool false (is not valid)

Back up to _get_collection_playbook_path
&nbsp;&nbsp;&nbsp;&nbsp;Nothing is returned because it returns false
&nbsp;&nbsp;&nbsp;&nbsp;fyi forcing it to return true on this line ( acr = AnsibleCollectionRef.try_parse_fqcr(playbook, u'playbook')) also errors out another way.

### Code Block 1

    def is_valid_fqcr(ref, ref_type=None):
        """
        Validates if is string is a well-formed fully-qualified collection reference (does not look up the collection itself)
        :param ref: candidate collection reference to validate (a valid ref is of the form 'ns.coll.resource' or 'ns.coll.subdir1.subdir2.resource')
        :param ref_type: optional reference type to enable deeper validation, eg 'module', 'role', 'doc_fragment'
        :return: True if the collection ref passed is well-formed, False otherwise
        """

        ref = to_text(ref)
        if not ref_type:
            return bool(re.match(AnsibleCollectionRef.VALID_FQCR_RE, ref))

        return bool(AnsibleCollectionRef.try_parse_fqcr(ref, ref_type))

### Code Block 2

I would think once you fix the recursion issue this line of code would be changed to include the ref_type.

if not AnsibleCollectionRef.is_valid_fqcr(ref): >> if not AnsibleCollectionRef.is_valid_fqcr(ref, ref_type):

### Actual Results

```console
ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/nginxinc.nginx_core.playbooks.deploy-nginx.yml' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/import_playbook.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/import_playbook.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
I believe the problem here is that your playbook contains a hyphen, which is not allowed in collection resources.  Collection resources must adhere to python naming requirements.

So renaming the playbook to `deploy_nginx.yml` instead of `deploy-nginx.yml` should allow you to use it, which of course requires using an underscore in your `import_playbook` too.

**Comment by provenvelocity** [comment]
That is inside the collection I'm accessing from NGINX Inc., not in my code.

It also does not work when I try any of the other test playbooks that are in my test code that are one word.

**Comment by sivel** [comment]
ansible-core 2.11 is the first version to provide the feature, so it's possible authors may have included playbooks that would not work before we actually supported them.

I have tested that properly named playbooks do work, and confirmed that we have tests that assert that they work:

https://github.com/ansible/ansible/blob/devel/test/integration/targets/collections/import_collection_pb.yml

The nginx collection authors will need to rename their playbooks to be functional for users.

**Comment by provenvelocity** [comment]
I get the same error, but now that I have the debugging I see that it finds it once the name is correct but it still gets the same error. 

I will follow the debug path to see why it fails. 
Also the code I outlined still has a potential recursion in it :)

new debug statement showing it it was found
0.1 AnsibleCollectionRef(collection='provenvelocity.devops_test', subdirs='', resource='main')

But the same error.

INFO     Running default > converge
ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/provenvelocity.devops_test.main' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option

**Comment by provenvelocity** [comment]
Also, I picked the Nginx stuff at random to just give me an outside testing solution because I could not get the local repo to work with testing.

**Comment by sivel** [comment]
Here is an example, that shows it working as expected:

```
├── 74917.yml
│   │ - import_playbook: foo.bar.baz_qux.yml
└── collections
    └── ansible_collections
        └── foo
            └── bar
                └── playbooks
                    └── baz_qux.yml
                        │ - hosts: localhost
                        │   gather_facts: false
                        │   tasks:
                        │     - debug:
```

```
$ ansible-playbook 74917.yml

PLAY [localhost] *****************************************************************************************************************************************************************************************************************************

TASK [debug] *********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "Hello world!"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

I see you are using molecule.  You should remove that from your reproducer, as it could be a bug in molecule itself, and not within ansible-core.

**Comment by provenvelocity** [comment]
I will but the pattern I'm using is to get my whole company testing their ansible code more so it is tightly tied to molecule.  The collection is located just fine if I run a role from the collection using molecule but not if I import the playbook from the collection.  All molecule is doing is running the command below which should be similar to the command you use in your testing. 

ansible-playbook --inventory /Users/H217689/.cache/molecule/devops_test/default/inventory --skip-tags molecule-notest,notest /Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/converge.yml


I have a repo that reproduces the issue end to end.  I will try wright a simple playbook outside of molecule that will add to the reproduction step to finds out where things are falling down.

The import_module below is not returning the collection path.

def _get_collection_playbook_path(playbook):

    acr = AnsibleCollectionRef.try_parse_fqcr(playbook, u'playbook')
    print ('0.1',acr)
    if acr:
        try:
            # get_collection_path
            pkg = import_module(acr.n_python_collection_package_name)
            print ('0.1.1',acr)

**Comment by provenvelocity** [comment]
I copied the inventory and converge file from molecule and took molecule itself out and it did find the needed playbook and import it.

UGGGG LOL thanks

I have been having no luck with them responding to tickets and just closing bugs... do you know anyone on that team? 

Also, there is still the recursion issue in the code FYI. The code will have a recursion if anyone ever calls is_valid_fqcr with a ref_type.

**Comment by sivel** [comment]
molecule is a community maintained project that we don't really have any interaction with.

**Comment by provenvelocity** [comment]
They will be working on a fix now thanks. 

May I make one suggestion that if ansible sees that its current folder struct is a galaxy collection then it loads that collection much the way you add roles folder to role path no matter what.

**Comment by provenvelocity** [comment]
Then my code would have worked… but it did not until I fixed molecule you
force it to include the current collection in the collections path…. :)

**Comment by bcoca** [comment]
I use this to develop/maintain collections, but we also have many
tests for it in CI , like the following one
https://github.com/ansible/ansible/blob/devel/test/integration/targets/collections/posix.yml#L14



-- 
----------
Brian Coca

**Comment by provenvelocity** [comment]
I am still having issues. I have set up a test repo to reproduce. 
 
Adding this conversation because these people should know the code does not work in all scenarios. 
https://github.com/ansible/ansible/pull/67435

###Setup
git clone git@github.com:provenvelocity/devops_test.git
git fetch --all
git checkout Collections_issue 
python3 -m venv ~/.pv_devops/pv_devops_venv
source ~/.pv_devops/pv_devops_venv/bin/activate
pip install --upgrade pip
pip install -r requirements_withlock.txt

###to reproduce the Ansible Collections issue 

branch: Collections_issue 

####Run:

ansible-playbook ./molecule/default/converge.yml -i localhost

####Wrong Result:
```
└─▪ (pv_devops_venv) ansible-playbook ./molecule/default/converge.yml -i localhost
[WARNING]: Unable to parse /Users/H217689/kp/ansible/provenvelocity/devops_test/localhost as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/provenvelocity.devops_test.palybooks.main.yml' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option
```

####Core Issue:
Ansible does not see the current collection as a collection. When using a collection to run a playbook you should not have to copy it to an anssible_collections folder to ensure ansible sees it that would mean you same code in two locations even though the working folder is the current folder and not the remote folder. This makes it overly complicated for the user to use collections.

In the roles_path var, you add ":roles" to is you should add the current folder to collections_path var if it has a galaxy.yml in it. This will fix the issue.

You can see how not doing this in many other tools causes issues in other tools because the pattern of collections is unique and not how roles or playbooks are normally treated.  Such as having to have it on a special path... This is fine for dependencies but not for the current collection repo your in. 

###Further Issues and debugging:
"
####Also:

- Running it like this allows you to see (some times)(some times not) the types ansible views the files as. You can see that only some are playbooks and some are not even though they are. 
- Also it thinks this is a file but it is not there. "molecule/provenvelocity.devops_test.main.yml"... These are all other issues but first I'm tackling the collections issue.
- Ansible lint uses ansible core for this so this is still an ansible issue BTW. ( I think :))

```
─▪ (pv_devops_venv) ansible-lint -vvv
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(cache_dir='/Users/H217689/.cache/ansible-lint/5e737a', colored=True, configured=True, cwd=PosixPath('/Users/H217689/kp/ansible/provenvelocity/devops_test'), display_relative_path=True, exclude_paths=['/Users/H217689/kp/ansible/provenvelocity/devops_test/.cache', '/Users/H217689/kp/ansible/provenvelocity/devops_test/.github'], lintables=[], listrules=False, listtags=False, parseable=False, parseable_severity=False, quiet=False, rulesdirs=['/Users/H217689/.pv_devops/pv_devops_venv/lib/python3.9/site-packages/ansiblelint/rules'], skip_list=['204', 'yaml'], tags=[], verbosity=3, warn_list=['experimental', 'role-name'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'requirements': '**/meta/requirements.yml'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.yml'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], mock_modules=[], mock_roles=[], loop_var_prefix=None, var_naming_pattern='^[a-z_][a-z0-9_]*$', offline=False, project_dir='.', extra_vars=None, enable_list=[], skip_action_validation=True, rules={}, format='rich', progressive=False, rulesdir=[], use_default_rules=True, config_file='/Users/H217689/kp/ansible/provenvelocity/devops_test/.ansible-lint', version=False)
INFO     Running ansible-galaxy role install --roles-path /Users/H217689/.cache/ansible-lint/5e737a/roles -vr requirements.yml
INFO     Running ansible-galaxy collection install -p /Users/H217689/.cache/ansible-lint/5e737a/collections -vr requirements.yml
INFO     Added ANSIBLE_COLLECTIONS_PATH=/Users/H217689/.cache/ansible-lint/5e737a/collections
DEBUG    Loading rules from /Users/H217689/.pv_devops/pv_devops_venv/lib/python3.9/site-packages/ansiblelint/rules
INFO     Running ansible-galaxy role install --roles-path /Users/H217689/.cache/ansible-lint/5e737a/roles -vr requirements.yml
INFO     Running ansible-galaxy collection install -p /Users/H217689/.cache/ansible-lint/5e737a/collections -vr requirements.yml
INFO     Added ANSIBLE_COLLECTIONS_PATH=/Users/H217689/.cache/ansible-lint/5e737a/collections:/Users/H217689/.cache/ansible-lint/5e737a/collections
Loading custom .yamllint config file, this extends our internal yamllint config.
INFO     Discovered files to lint using: git ls-files -z
INFO     Executing syntax check on molecule/docker/verify.yml (0.83s)
INFO     Executing syntax check on molecule/default/verify.yml (0.84s)
INFO     Executing syntax check on molecule/docker/converge.yml (0.86s)
INFO     Executing syntax check on molecule/default/converge.yml (0.88s)
INFO     Executing syntax check on playbooks/main.yml (0.88s)
DEBUG    Examining tests/test.yml of type yaml
DEBUG    Examining tests/shared_tests.yml of type yaml
DEBUG    Examining molecule/default/converge.yml of type playbook
DEBUG    Examining molecule/docker/verify.yml of type playbook
DEBUG    Examining requirements.yml of type requirements
DEBUG    Examining .yamllint of type yaml
DEBUG    Examining .env.yml of type yaml
DEBUG    Examining molecule/docker/converge.yml of type playbook
DEBUG    Examining molecule/default/molecule.yml of type yaml
DEBUG    Examining playbooks/main.yml of type playbook
DEBUG    Examining molecule/default/verify.yml of type playbook
DEBUG    Examining main.yml of type yaml
DEBUG    Examining galaxy.yml of type galaxy
DEBUG    Examining molecule/docker/molecule.yml of type yaml
DEBUG    Examining .ansible-lint of type yaml
DEBUG    Examining molecule/provenvelocity.devops_test.main.yml of type playbook
DEBUG    Examining tests/shared_tests.yml of type playbook
WARNING  Listing 1 violation(s) that are fatal
load-failure: [Errno 2] No such file or directory: 'molecule/provenvelocity.devops_test.main.yml' (filenotfounderror)
molecule/provenvelocity.devops_test.main.yml:1
```
####Also:

If you notice in the example above it does not determine that main.yml in the root of the collection is a playbook so it marks it YAML type I then use the same file in the playbooks folder and then get the error below which is the same as the first error and explained below. 

Ansible lint uses ansible core for this so this is still an ansible issue BTW. ( I think :))

```
└─▪ (pv_devops_venv) ansible-lint -vvv
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(cache_dir='/Users/H217689/.cache/ansible-lint/5e737a', colored=True, configured=True, cwd=PosixPath('/Users/H217689/kp/ansible/provenvelocity/devops_test'), display_relative_path=True, exclude_paths=['/Users/H217689/kp/ansible/provenvelocity/devops_test/.cache', '/Users/H217689/kp/ansible/provenvelocity/devops_test/.github'], lintables=[], listrules=False, listtags=False, parseable=False, parseable_severity=False, quiet=False, rulesdirs=['/Users/H217689/.pv_devops/pv_devops_venv/lib/python3.9/site-packages/ansiblelint/rules'], skip_list=['204', 'yaml'], tags=[], verbosity=3, warn_list=['experimental', 'role-name'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'requirements': '**/meta/requirements.yml'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.yml'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], mock_modules=[], mock_roles=[], loop_var_prefix=None, var_naming_pattern='^[a-z_][a-z0-9_]*$', offline=False, project_dir='.', extra_vars=None, enable_list=[], skip_action_validation=True, rules={}, format='rich', progressive=False, rulesdir=[], use_default_rules=True, config_file='/Users/H217689/kp/ansible/provenvelocity/devops_test/.ansible-lint', version=False)
INFO     Running ansible-galaxy role install --roles-path /Users/H217689/.cache/ansible-lint/5e737a/roles -vr requirements.yml
INFO     Running ansible-galaxy collection install -p /Users/H217689/.cache/ansible-lint/5e737a/collections -vr requirements.yml
INFO     Added ANSIBLE_COLLECTIONS_PATH=/Users/H217689/.cache/ansible-lint/5e737a/collections
DEBUG    Loading rules from /Users/H217689/.pv_devops/pv_devops_venv/lib/python3.9/site-packages/ansiblelint/rules
INFO     Running ansible-galaxy role install --roles-path /Users/H217689/.cache/ansible-lint/5e737a/roles -vr requirements.yml
INFO     Running ansible-galaxy collection install -p /Users/H217689/.cache/ansible-lint/5e737a/collections -vr requirements.yml
INFO     Added ANSIBLE_COLLECTIONS_PATH=/Users/H217689/.cache/ansible-lint/5e737a/collections:/Users/H217689/.cache/ansible-lint/5e737a/collections
Loading custom .yamllint config file, this extends our internal yamllint config.
INFO     Discovered files to lint using: git ls-files -z
INFO     Executing syntax check on molecule/default/converge.yml (0.70s)
INFO     Executing syntax check on molecule/docker/verify.yml (0.82s)
INFO     Executing syntax check on molecule/default/verify.yml (0.83s)
INFO     Executing syntax check on playbooks/main.yml (0.86s)
INFO     Executing syntax check on molecule/docker/converge.yml (0.86s)
WARNING  Listing 1 violation(s) that are fatal
internal-error: Unexpected error code 1 from execution of: ansible-playbook --syntax-check molecule/default/converge.yml
molecule/default/converge.yml:1 [WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/provenvelocity.devops_test.playbooks.main.yml' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option
```

####Expected Result:
Converge and install playbook.

####Issues:
- Ansible does not see the current collection as a collection unless it is in their collections folders. Currently, ansible lint does not place them there so this command fails. With role_paths you add ":role" to the paths you need to add the current folder if it has a galaxy.yml file in it. So you do not need to copy the current collection outside of the path it is in to see it in the collections struct.
- If you run this with ansible lint you will see that it sets the path of a collection and copies the collections there. This gets you past one issue but then you get another set of issues. Not all Types are respected and when it runs internally ansible-playbook command you get back to the core issue which is ansible does not see the current collection and in this example is not respecting the ANSIBLE_COLLECTION_PATH when run in a nested way as out of ansible-lint. (This may actually be an issue with ansible lint not passing the configure ENV to the command it runs to test the code. )
- This also happens in molecule because it runs the code from ansible-lint.

####Expected Result:
Converge and install playbook.

####Issues:
- Ansible does not see the current collection as a collection unless it is in their collections folders. Currently, ansible lint does not place them there so this command fails. With role_paths you add ":role" to the paths you need to add the current folder if it has a galaxy.yml file in it. So you do not need to copy the current collection outside of the path it is in to see it in the collections struct.
- If you run this with ansible lint you will see that it sets the path of a collection and copies the collections there. This gets you past one issue but then you get another set of issues. Not all Types are respected and when it runs internally ansible-playbook command you get back to the core issue which is ansible does not see the current collection and in this example is not respecting the ANSIBLE_COLLECTION_PATH when run in a nested way such as out of ansible -lint. (This may actually be an issue with ansible lint not passing the configure ENV to the command it runs to test the code. )
- This also happens in molecule because it runs the code from ansible-lint.

**Comment by sivel** [comment]
```
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/provenvelocity.devops_test.palybooks.main.yml' on the Ansible Controller.
```

`palybooks` != `playbooks`

**Comment by provenvelocity** [comment]
Sorry im dislexic... Same issue

└─▪ (pv_devops_venv) ansible-playbook ./molecule/default/converge.yml -i localhost
[WARNING]: Unable to parse /Users/H217689/kp/ansible/provenvelocity/devops_test/localhost as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
ERROR! Unable to retrieve file contents
Could not find or access '/Users/H217689/kp/ansible/provenvelocity/devops_test/molecule/default/provenvelocity.devops_test.playbooks.main.yml' on the Ansible Controller.
If you are using a module and expect the file to exist on the remote, see the remote_src option

**Comment by provenvelocity** [comment]
@mkrizek can you reopen for triage 

Thanks

**Comment by sivel** [comment]
Also, you will need to configure your path to the collection, unless you `git clone` the collection into one of the default dirs.  Otherwise ansible has no idea about the collection your are trying to use.

https://docs.ansible.com/ansible/latest/reference_appendices/config.html#collections-paths

Something like:

```
ANSIBLE_COLLECTIONS_PATH=`pwd` ansible-playbook ./molecule/default/converge.yml -i localhost
```

**Comment by provenvelocity** [comment]
Ya That is why like the molecule team I used ansible-lint to test and manage collections testing. Ansible-lint does set the collections path explicitly and it also has the ansible_collections folder in the path and it does it great.... but as you can see in my examples it even has the same problem. Because it calls ansible-playbook and loses the env setup. 

Per the documentation, this should be possible so we need to change the documentation or fix the code to simply add in the current collection to the collection path. 

In my testing it's a one-line fix in ansible-lint and molecule... to do this, but I have not looked at ansible.

**Comment by provenvelocity** [comment]
@webknjaz @bcoca  I did a detailed writeup of the core issue could you take a look also and weigh in :)

**Comment by provenvelocity** [comment]
BTW I monkey patched this with one line ansible-lint to allow for a custom hardcoded cache dir and then setting that cache dir in every tool and location possible and it all worked.  It's a super hacky fix but it worked.... I'm going to look at ansible code and see if I can do the same, though again I'm trying to avoid monkey patching.

**Comment by sivel** [comment]
Explicit invocations of a path based playbook within a collection will not auto add the collection path to the search lookup.

While this may be something lint and molecule will do, it is not something ansible itself will add a feature for.

**Comment by sivel** [comment]
I'm not sure what you are saying then.  If you want to use a collection, and have ansible use it, you must properly define the path if in a non-standard location, or put it in a pre-defined path.  Testing frameworks can decide to short circuit some of this for you, which ansible-test does.  We do not special case running path based playbooks from a collection and do anything special to enable that collection to be inserted into the collection search path.

Sounds like you are asking for what I just said we have no intention of implementing.

**Comment by provenvelocity** [comment]
If you are in a collection you have to and you want to test you have to then copy it to a special folder. An Artificially constrained and defined folder that has no programmatic need other than to handle dependencies every time you want to test a collection.

 This makes no sense from a dev or ci point of view. 
 
Also any tool that then calls ansible in its testing has to explicitly make sure that the env and collections path which you can change is explicitly defined which then also falls apart. Because you have to make sure it is passed from the current running state to the python command call of ansible-playbook or it also fails. 

ansible-lint, molecule, ansible-test and I'm sure many many many other tools do just that and all fall apart if you use a custom collections path and or are testing the current collection your in. 

All of these issues are fixed if you add the current folder in the path (which the roles team does) and separately add the dependency resolution of collections to ansible-playbook which is just good practice to make sure the dependencies are correct. If you do not at least add the current folder then even basic adoption and testing of collections does not work. 

The reason tools set the collections path is there are tooooooooo many options for collections paths and if you have one version in one and another in another folder this just gets messier in which one overrides the other or you missed one and the code works local but not on the server because you have a dependency resolved in some folder you did not know.... It is all just a very very bad pattern and falls apart the more I'm trying to adopt it because as ansible states this is the future struct of the code.

**Comment by sivel** [comment]
As stated above, this is not something we have intention of implementing.  If you are calling a path based playbook, regardless of the fact that it resides in a directory that may or may not look like a collection, you must still place the collection in proper pathing, and configure ansible to utilize that path if it is in a non-standard location.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.libera.chat
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by provenvelocity** [comment]
Every tool that calls ansible-playbook in python and sets a custom. collections path prior to that or is in a collection will fail period  EVERY TOOL....

**Comment by provenvelocity** [comment]
I will keep this short since I don't think you actually even read my posts.... 

Running the following command also does not include the current collection you are in... which means the only way to move it over is manually EVERY TIME I TEST or use my upgraded ansible-lint that now can better handle collections... because they already had a hey copy current collection over because well ansible does not listen to customers... 

 ansible-galaxy collection install -p ./collections -vr requirements.yml
 
 If you try to reference the local collection you get a nasty recursion effect....

**Comment by bcoca** [comment]
ansible-playbook is NOT a develompent tool but a runtime one, having it auto detect a possible collection wrapping the current playbook and then auto inject that into the 'available collections' is a bad idea. 

As convenient as that is for development, it creates many issues in a production context.

**Comment by provenvelocity** [comment]
You develop code in a framework and then run it with ansible-playbook yes,
but in doing so it does all sorts of things to ensure it can operate and
the operational choices you made are now effecting your customers.

The ansible engineering team made the choice to build and require
collections to operate in this manner. This is the first pattern in ansible
that now requires further steps to operate.

So this code pattern will not work without steps outside of
ansible-playbook to run means that by default you are 100% tied to these
steps already. Your customers did not do this you already did….

All we are asking is as a company you must handle for the ill thought out
end result of this choice. If ansible-playbook fundamentally will not work
without a side development process then the choice has been made for you.
And to not automate and or handle for this or change in direction is  bad
for business.

**Comment by bcoca** [comment]
This is not new to Ansible nor exclusive to collections, roles/plugins/modules all had to be installed in either configured paths or following an adjacent path convention. 

While we always strive to make the workflows easier and simpler for users, we also have to weigh into other considerations like security, scalability and reliability (which users also demand).  

We do have many plans and discussions around making content development easier, some is based on documenting use workflows, others on developing and/or enhancing existing development tools.  What does not make sense in this context is to bake it into the ansible-playbook.

**Comment by provenvelocity** [comment]
To be clear with playbooks, plugins, modules, and roles you do not require
that I know of a user to copy them to a folder before ansible-playbook can
see them. That is is what I'm saying.,... That ansible-playbook is not
aware that it is in a collection that is operating in a specific context.
Fine, you want separation of duties ok in this case not my cup of tee but
at least make it aware that it will fail if some basic requirements are
met.

I am trying to help so let me break down what this means to a customer:


   1. Right now you have to run ansibl-galixy to install your
   dependencies great not that hard... but ansible-galaxy does not handle for
   the fact that you are in a collection and has a recursion error if you tell
   it you are. So then you have to manually copy over your code every time you
   want to test your code prior to release.
   2. Every time I make a change and want to test it I have to
   manually copy over code to a special folder because ansibl-playbook is not
   aware of the context it is running in. Ok Not fun but hey this is what
   ansible is telling us is the future of their code framework.
   3. Then once I'm done because I can't release the code outside of my
   company and the internal galaxy is completely unusable because of bad UI
   and RBAC model...
   4. I then have to recode all my pipelines to handle for running galaxy,
   and then running a manual step on top of that.
   5. I then have to code a fix for shared pipelines because ansible copies
   all the collections to the same location on the same agent across teams
   without versioning the folders so they overwrite them... but many of the
   tools struggle with a custom collection path because of the same issues
   noted above.
   6. Once that is done I should be able to at least test the code but
   Ansible-lint and molecule do not work correctly because of the same said
   issues as said above soooo I then have to recode and fork them and wait for
   them to accept the fix and run off forked code.... ok more complexity but
   ok lets do this.
   7. Then Once I get this all working sense my company just bought tower
   and we are quite large and converting quite a bit of code to anible I need
   this all to work in tower.
   8. I then work through the same issue in tower and sense we are not
   dynamic enough to have ephemeral instances in tower many of the teams are
   running on the same instance group... I now foresee another issue with this
   dependency cache as I had with the testing pipeline.

This is the barrier of entry to using collections as they stand now... This
is the issue with not making ansible-playbook contextually aware of the
type of playbook it is running and adding it to the internal list of
collections. And also made aware of the fact that now its code framework
has dependencies and will not work if you don't handle for that.

All this complexity is what I have had to tackle to try and use the latest
code model from ansibel. That is not why we switched to ansible....

**Comment by bcoca** [comment]
> To be clear with playbooks, plugins, modules, and roles you do not require
that I know of a user to copy them to a folder before ansible-playbook can
see them. That is is what I'm saying.,... That ansible-playbook is not
aware that it is in a collection that is operating in a specific context.
Fine, you want separation of duties ok in this case not my cup of tee but
at least make it aware that it will fail if some basic requirements are not
met.

This is just not true, plugins, modules and roles will not be found unless they are in the expected paths. 
The playbooks (like inventory sources or extra vars files) are one of the few things that pare passed as full paths (arguments) to the tools.

>1 
This is true for roles also, for modules/plugins there was no good way to install, only manually or indirectly via roles
As for the recursion error, I'm not sure what is happening, please open a ticket with the specific issues and errors.
>2
No, this depends on the workflow you setup, I develop collections and can test them w/o having to copy anything, there are many ways to achieve this, either by using configuration (file or env vars) or symlinks to enable a simpler flow.
>3 
I'm not sure what the problem is here, but there are many options for releasing the code, many self hosted, a tarball, shared filesystem with installed collection, git repo to more complex options like setting up your own galaxy or automation hub.
>4 
This is one way to do it, personally I don't see much of an issue with ansible-galaxy running in CI ..I, like many others have done this for roles for a long time, again, this is not really a requirement unless you depend on external roles/collections and always want the latest.  Otherwise you can simply go against the ones you already marked stable and have installed.
>5
I'm not sure why you need this to be the case, but this is also how roles have always worked, there was never a 'versioned'  setup for roles either (which i did propose at one time https://github.com/ansible/proposals/issues/23). So it is not a new issue introduced by collections. There are also several ways to work around this, but requires a lot more context.
>6
I'm aware of a few issues with lint and molecule with collections and I know the maintainers have been working on them, a lot of them have already been solved, nothing I know of is a showstopper or would require the feature you are asking for.
>7
possibly, tower has some additional restrictions on what and how to run things, but i'm not sure how this relates to the feature either
> 8
I don't see how this is a  problem, I probably need more info.

Sorry to hear, but I don't see that most of that complexity is inherit to Ansible, it seems a lot is derived from the context you are in.

**Comment by provenvelocity** [comment]
*This is just not true, plugins, modules and roles will not be found unless
they are in the expected paths.The playbooks (like inventory sources or
extra vars files) are one of the few things that pare passed as full paths
(arguments) to the tools.*

   - A location that ansibl-playbook is aware of inherently and does not
   handle outside of the code.... Ansible is made aware of these but not the
   collection it is in....

*Sorry to hear, but I don't see that most of that complexity is inherit to
Ansible, it seems a lot is derived from the context you are in.*

   - None of this complexity is from me but from the new framework from
   ansible... collections. The context is ansible collections.

*No, this depends on the workflow you setup, I develop collections and can
test them w/o having to copy anything, there are many ways to achieve this,
either by using configuration (file or env vars) or symlinks to enable a
simpler flow.*

   - I did collection init and then added dependency and nothing in
   ansible moves the current collection to any collection folder. This will be
   the same in any pipeline I develop especially.... Also, I support over 200+
   dev teams and so can you imagine me having to educate and support and get
   them all to understand these steps to develop their new ansible code. I'm
   trying to make it seamless...

The rest have similar issues and as you said its all about context and this
context is all ansible collections and how they function in a real world
company and not just in a small dev team like ansible.  We will have over
2-300 people wrighting ansible collections and consuming internal
collections for their teams for example.

**Comment by bcoca** [comment]
>   - A location that ansibl-playbook is aware of inherently and does not
   handle outside of the code.... Ansible is made aware of these but not the
   collection it is in....

A role is not aware of the role it inside nor playbook it is used in unless being executed from that playbook, even then the  role must be in the expected paths or playbook cannot access it either. The same is true for modules and plugins

The only 'awareness' a playbook has is what we call 'adjacent'  directories, like having roles/ for roles, library for modules/in the same directory, having collecitons/ansbile_collections/  also works, this is no  different than previous methods (except requiring a 2nd directory). This is also why we added --playbook-dir to many utilities so not only ansible-playbook can use this concept. 

> None of this complexity is from me but from the new framework from
   ansible... collections. The context is ansible collections.

I don't know if we are communicating well here, but that is not what I'm inferring from your examples or statements, many of which are just incorrect.

>  I did collection init and then added dependency and nothing in
   ansible moves the current collection to any collection folder. This will be
   the same in any pipeline I develop especially.... Also, I support over 200+
   dev teams and so can you imagine me having to educate and support and get
   them all to understand these steps to develop their new ansible code. I'm
   trying to make it seamless...

The  collection init function allows you to do things many ways, one of them is by creating it in the 'configured directory', this is a CHOICE you make, not an imposition.  As for why this choice exists, in some contexts developers rather not automatically initialize a collection in the place it would be picked up at runtime, until they decide it is ready for such a thing.
An easy way to list configured paths is to use `ansbile-config dump |grep COLLECTIONS_PATHS`.

The collection is not designed for a small team, quite the contrary, they were made to deal with the explosion of content that could not be handled anymore in the core project and to enable multiple teams to work together and minimize conflict (hence the namespaces).  It might not work with the workflow you have in mind, but that is probably true of many things, nothing works with 100% of everything.

If you have specific suggestions were are open to hearing them, that does not mean we will always implement them, there are many things we need to consider. I understand why you want the 'contextual collection autoload' and have given you the reasons why we don't think it is a good idea in ansible-playbook. ansilbe-test/lint/molecule are actually implementing such things.

You bring up other issues but either don't supply enough information (a detailed ticket on each would be helpful) or just make incorrect statements that make it hard for me to accept the issue.

**Comment by provenvelocity** [comment]
I am leary of even trying to make any more tickets…. I have spent way to
much of my day job trying to explain a simple issue with the implementation
of the doucmention you have for collections. I did what it said and it did
not and does not work as documented.

I have several other tickets open and will add one for the Glaxy once I
have time. I’m currently working downstream to fix ansible lint and
molecule so I can release the framework to the company and they can work to
convert to ansible all using the same frame work.

I know collection was made to solve big problems that is why I’m using it
but your not exposed to the scale and institutional bs I am and have to
deal with and all I’m trying to do is educate the eng team on the
implementation issues of the current model for collections.

I have been on the advisory board of Jenkins, hashicorp, chef, puppet… and
have been doing this type of integrations for 25 plus years. Trust me the
issues I have stated will be an issue for any large enterprise to adopt
collections…

**Comment by thaumos** [comment]
Hi @provenvelocity, 

I have also reached out to you via the ticket you opened with us at Red Hat to speak about this.  Look for my updates there, please.  Speak to you soon.
